### PR TITLE
Vasilev Ivan lab3 option 1

### PR DESCRIPTION
**Pull request by AdAndRoll**

### CODE REVIEW

**Review by aobolensk — APPROVED** [review]

File: llvm/test/lab3/vasilev_ivan/test.mir [hunk 1]
```diff
@@ -0,0 +1,495 @@
+# RUN: llc -mtriple x86_64-unknown-linux-gnu --load=%llvmshlibdir/X86VasilevMulAddPass%shlibext -run-pass=x86-Vasilevmuladd  %s -o - | FileCheck %s
+
+--- |
+  ; ModuleID = '/home/adandroll/lab3/llvm-nnsu-2024/llvm/test/lab3/vasilev_ivan/test.ll'
```
Comment (m-ly4) [inline llvm/test/lab3/vasilev_ivan/test.mir]: Could you please add the following test:
```
__m128d testX(__m128d a, __m128d b) {
    auto tmp = a * b;
    return tmp + tmp;
}
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/vasilev_ivan/test2.mir [hunk 2]
```diff
@@ -0,0 +1,560 @@
+# RUN: llc -mtriple x86_64-unknown-linux-gnu --load=%llvmshlibdir/X86VasilevMulAddPass%shlibext -run-pass=x86-Vasilevmuladd  %s -o - | FileCheck %s
+
+--- |
+  ; ModuleID = '/home/adandroll/lab3/llvm-nnsu-2024/llvm/test/lab3/vasilev_ivan/test2.ll'
+  source_filename = "/home/adandroll/lab3/llvm-nnsu-2024/llvm/test/lab3/vasilev_ivan/test2.cpp"
+  target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+  target triple = "x86_64-pc-linux-gnu"
+  
+  ; Function Attrs: mustprogress noinline nounwind optnone uwtable
+  define dso_local noundef <2 x double> @_Z12muladd_test1Dv2_dS_S_(<2 x double> noundef %0, <2 x double> noundef %1, <2 x double> noundef %2) #0 {
```
Comment (m-ly4) [inline llvm/test/lab3/vasilev_ivan/test2.mir:10]: Please remove duplicated test cases.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/test/lab3/vasilev_ivan/test2.mir (hunk 2 above)
Comment (AdAndRoll) [reply llvm/test/lab3/vasilev_ivan/test2.mir:10]: yes, sorry, I just wanted to make 1 file, so I moved all the tests + added a new one, forgot to delete the previous file with the tests

Comment (AdAndRoll) [reply llvm/test/lab3/vasilev_ivan/test2.mir:10]: maybe just delete test.mir?

File: llvm/test/lab3/vasilev_ivan/test2.mir [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+body:             |
+  bb.0 (%ir-block.3):
+    liveins: $xmm0, $xmm1, $xmm2
+  
+    frame-setup PUSH64r killed $rbp, implicit-def $rsp, implicit $rsp
+    CFI_INSTRUCTION def_cfa_offset 16
+    CFI_INSTRUCTION offset $rbp, -16
+    $rbp = frame-setup MOV64rr $rsp
+    CFI_INSTRUCTION def_cfa_register $rbp
+    ; CHECK: MOVAPDmr $rbp, 1, $noreg, -64, $noreg, killed renamable $xmm0 :: (store (s128) into %ir.4)
```
Comment (m-ly4) [inline llvm/test/lab3/vasilev_ivan/test2.mir]: As I see in description of 3rd test two fmuladd instructions are expected to be made, but I don't see them. It seems that it is due to using MIR generated after regalloc and stack allocation. Could you use MIR generated earlier for simpler input?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — CHANGES_REQUESTED** [review]
There is still a remark to be fixed.

**Comment by AdAndRoll** [comment]
sorry, I don't quite understand what remark we are talking about, Tests is added, at the end of the test2.mir file, two checks of the third test, noted earlier

**Comment by m-ly4** [comment]
> sorry, I don't quite understand what remark we are talking about, Tests is added, at the end of the test2.mir file, two checks of the third test, noted earlier

https://github.com/NN-complr-tech/llvm-nnsu-2024/pull/151/files#r1605804446

I expect fmuladd instructions here.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**